middle_end/flambda/simplify/simplify_apply_expr.rec.ml

```diff
@@ -637,20 +637,22 @@ let simplify_function_call dacc apply ~callee_ty
     down_to_up dacc ~rebuild:Simplify_common.rebuild_invalid
 
 let simplify_apply_shared dacc apply : _ Or_bottom.t =
-  let min_name_mode = Name_mode.normal in
-  match S.simplify_simple dacc (Apply.callee apply) ~min_name_mode with
-  | Bottom, _ty -> Bottom
-  | Ok callee, callee_ty ->
-    match S.simplify_simples dacc (Apply.args apply) ~min_name_mode with
-    | _, Bottom -> Bottom
-    | _changed, Ok args_with_types ->
-      let args, arg_types = List.split args_with_types in
+  let callee_ty =
+    S.simplify_simple dacc (Apply.callee apply) ~min_name_mode:NM.normal
+  in
+  if T.is_obviously_bottom callee_ty then Bottom
```

Comment:
Here (and in a few other places) I believe that the type returned by `simplify_simple` will never be obviously bottom. Even when the environment has a bottom equation associated to the given simple, `Typing_env.type_simple_in_term_exn` will return an alias t the original simple and not a bottom type.
So I would be in favour of removing these branches (as far as I understand, the `Or_bottom` result from the previous version would only return `Bottom` when merging the rec_info would fail).

Reply:
It seems like we should have some detection of `Bottom` here though, no?  (It seems like that wasn't there before either, for the reason you mention.)  Do you have a preferred means of doing that?

Reply:
I've already said that I wouldn't mind forbidding Bottom equations in the environment (we never generate any of those at the moment anyway).
Even without going that far, I'm struggling to see how we could end up in a situation where the type for a simple is Bottom, but we didn't notice it earlier.
However, if we start introducing poisoned values (like @chambart suggested), then it would make sense to detect these here.

Reply:
Should be fixed in the latest commit.